### PR TITLE
Dont use row block field block offsets if not needed

### PR DESCRIPTION
**Pull request by radek-starburst**
*  Add support for RowType to BenchmarkDataGenerator and benchmarks
* Do not use/serialize/deserialize fieldBlockOffset when not needed
  * When there are no null rows within RowBlock then it's not needed to store fieldBlockOffsets.
This improves performance and reduces network traffic as fieldBlockOffsets array is not serialized or deserialized.
```
Benchmark  (before/after)                      (nullChance)  Mode  Cnt  Score   Error  Units
(before)   BenchmarkBlockSerde.deserializeRow  0             avgt  60   17.348  ±      0.109  ns/op
(after)    BenchmarkBlockSerde.deserializeRow  0             avgt  60   15.205  ±      0.084  ns/op
(before)   BenchmarkBlockSerde.deserializeRow  .01           avgt  60   18.020  ±      0.104  ns/op
(after)    BenchmarkBlockSerde.deserializeRow  .01           avgt  60   18.279  ±      0.113  ns/op
(before)   BenchmarkBlockSerde.deserializeRow  .10           avgt  60   16.859  ±      0.140  ns/op
(after)    BenchmarkBlockSerde.deserializeRow  .10           avgt  60   16.494  ±      0.157  ns/op
(before)   BenchmarkBlockSerde.deserializeRow  .50           avgt  60   10.273  ±      0.044  ns/op
(after)    BenchmarkBlockSerde.deserializeRow  .50           avgt  60   10.491  ±      0.063  ns/op
(before)   BenchmarkBlockSerde.deserializeRow  .90           avgt  60   4.351   ±      0.020  ns/op
(after)    BenchmarkBlockSerde.deserializeRow  .90           avgt  60   4.311   ±      0.013  ns/op
(before)   BenchmarkBlockSerde.deserializeRow  .99           avgt  60   2.919   ±      0.027  ns/op
(after)    BenchmarkBlockSerde.deserializeRow  .99           avgt  60   2.851   ±      0.008  ns/op
(before)   BenchmarkBlockSerde.serializeRow    0             avgt  60   20.862  ±      0.117  ns/op
(after)    BenchmarkBlockSerde.serializeRow    0             avgt  60   16.941  ±      0.089  ns/op
(before)   BenchmarkBlockSerde.serializeRow    .01           avgt  60   21.588  ±      0.097  ns/op
(after)    BenchmarkBlockSerde.serializeRow    .01           avgt  60   20.045  ±      0.102  ns/op
(before)   BenchmarkBlockSerde.serializeRow    .10           avgt  60   20.667  ±      0.068  ns/op
(after)    BenchmarkBlockSerde.serializeRow    .10           avgt  60   19.448  ±      0.126  ns/op
(before)   BenchmarkBlockSerde.serializeRow    .50           avgt  60   13.217  ±      0.090  ns/op
(after)    BenchmarkBlockSerde.serializeRow    .50           avgt  60   11.839  ±      0.042  ns/op
(before)   BenchmarkBlockSerde.serializeRow    .90           avgt  60   7.404   ±      0.017  ns/op
(after)    BenchmarkBlockSerde.serializeRow    .90           avgt  60   5.615   ±      0.009  ns/op
(before)   BenchmarkBlockSerde.serializeRow    .99           avgt  60   5.517   ±      0.007  ns/op
(after)    BenchmarkBlockSerde.serializeRow    .99           avgt  60   3.534   ±      0.009  ns/op

```

### CODE REVIEW

**Comment by sopel39** [comment]
Is it really a draft? If you think it's ready make it non-draft?

File: core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java [hunk 1]
```diff
@@ -35,40 +35,55 @@ public String getName()
...
         }
 
-        sliceOutput.appendInt(positionCount);
-        for (int position = 0; position < positionCount + 1; position++) {
-            sliceOutput.writeInt(fieldBlockOffsets[offsetBase + position] - startFieldBlockOffset);
-        }
         EncoderUtil.encodeNullsAsBits(sliceOutput, block);
+
+        if (isNullPresent) {
+            for (int position = 0; position < positionCount + 1; position++) {
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java:64]: nit: most likely it makes sense to compute `fieldBlockOffsets[offsetBase + position] - startFieldBlockOffset` for all positions, then use `writeBytes(wrappedIntArray(...))` instead of appending `writeInt` positions by positions

See how other serializers work (e.g. `io.trino.spi.block.IntArrayBlockEncoding`

File: core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlock.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package io.trino.spi.block;
+
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertNull;
+
+public class TestRowBlock
```
Comment (sopel39) [inline core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlock.java:24]: there is already `io.trino.block.TestRowBlock` (add tests there)

File: core/trino-main/src/test/java/io/trino/block/BenchmarkRowBlock.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+import static java.util.concurrent.TimeUnit.MICROSECONDS;
+
+@SuppressWarnings("MethodMayBeStatic")
+@State(Scope.Thread)
+@OutputTimeUnit(MICROSECONDS)
+@Fork(1)
+@Warmup(iterations = 3)
+@Measurement(iterations = 5)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkRowBlock
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BenchmarkRowBlock.java:57]: would it be possible to add benchmark case to `io.trino.execution.buffer.BenchmarkBlockSerde` instead?

**Comment by sopel39** [comment]
mind automation (maven checks)

File: core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java [hunk 4]
```diff
@@ -38,6 +38,7 @@
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java:38]: Improve commit message:
```
When there are no null rows within RowBlock
then it's not needed to store fieldBlockOffsets.
This improves performance and reduces
network traffic as fieldBlockOffsets
array is not serialized or deserialized.
```

Also, attach before/after benchmark results to this commit message

File: core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java [hunk 5]
```diff
@@ -109,6 +109,7 @@ public final Block copyPositions(int[] positions, int offset, int length)
             if (fieldBlockPositionCount == length) {
                 // No nulls encountered, discard the null mask
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java:110]: update comment

File: core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java [hunk 6]
```diff
@@ -35,40 +37,50 @@ public String getName()
...
             blockEncodingSerde.writeBlock(sliceOutput, rowBlock.getRawFieldBlocks()[i].getRegion(startFieldBlockOffset, endFieldBlockOffset - startFieldBlockOffset));
         }
 
-        sliceOutput.appendInt(positionCount);
-        for (int position = 0; position < positionCount + 1; position++) {
-            sliceOutput.writeInt(fieldBlockOffsets[offsetBase + position] - startFieldBlockOffset);
-        }
         EncoderUtil.encodeNullsAsBits(sliceOutput, block);
+
+        if (isNullPresent && fieldBlockOffsets != null) {
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java:61]: use:
```
if (fieldBlockOffsets != null) {
  if (!isNullPresent) {
    throw new IllegalStateException("fieldBlockOffsets should not be present when there are no null positions");
  }
   ...
}
```

File: core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java [hunk 7]
```diff
@@ -35,40 +37,50 @@ public String getName()
...
+            int[] positions = IntStream.range(0, positionCount + 1).map(i -> fieldBlockOffsets[offsetBase + i] - startFieldBlockOffset).toArray();
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java:62]: streams are not good in tight loops. You can use `io.airlift.slice.Slices#wrappedIntArray(int[], int, int)` to point at specific offset without copying data

File: core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlock.java (hunk 2 above)
Comment (sopel39) [reply core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlock.java:24]: ping!

File: core/trino-main/src/test/java/io/trino/execution/buffer/BenchmarkBlockSerde.java [hunk 8]
```diff
@@ -245,6 +232,44 @@ else if (TINYINT.equals(type)) {
 
             setup(sliceOutput.slice(), pagesSerde, pages);
         }
+
+        private void writeValue(Type type, Object value, BlockBuilder blockBuilder)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/execution/buffer/BenchmarkBlockSerde.java:236]: nit: you could make this a separate commit

File: core/trino-main/src/test/java/io/trino/execution/buffer/BenchmarkBlockSerde.java [hunk 9]
```diff
@@ -369,6 +394,18 @@ public void setup()
...
     }
 
+    @State(Thread)
+    public static class RowTypeBenchmarkData
+            extends TypeBenchmarkData
+    {
+        @Setup
+        public void setup()
+        {
+            RowType type = RowType.anonymous(ImmutableList.of(VARCHAR, BIGINT, LONG_DECIMAL_TYPE));
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/execution/buffer/BenchmarkBlockSerde.java:404]: You can just encode/decode single bigint field within `RowBlock`. That would be even better because it would expose `RowBlock` overhead even more

**Review by skrzypo987 — COMMENTED** [review]
Added some comments, will look again tomoroow

**Review by skrzypo987 — APPROVED** [review]
LGTM % comments

File: core/trino-main/src/test/java/io/trino/execution/buffer/BenchmarkBlockSerde.java [hunk 10]
```diff
@@ -27,6 +27,7 @@
 import io.trino.spi.type.DecimalType;
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/execution/buffer/BenchmarkBlockSerde.java:27]: merge `Add serializeRow and deserializeRow to BenchmarkBlockSerde` and ` Add support for RowType to BenchmarkDataGenerator` commits together

File: core/trino-main/src/test/java/io/trino/execution/buffer/BenchmarkBlockSerde.java (hunk 8 above)
Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/execution/buffer/BenchmarkBlockSerde.java:236]: I mean extraction of `writeValue` could be a separate commit since it's pure refactor

File: core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java [hunk 11]
```diff
@@ -35,40 +35,56 @@ public String getName()
...
-        }
         EncoderUtil.encodeNullsAsBits(sliceOutput, block);
+
+        if ((rowBlock.getRowIsNull() == null) != (fieldBlockOffsets == null)) {
+            throw new IllegalArgumentException("When rowIsNull is (non) null then fieldBlockOffsets should be (non) null as well");
+        }
+
+        if (fieldBlockOffsets != null) {
+            int[] positions = new int[positionCount + 1];
+            for (int position = 0; position < positionCount + 1; position++) {
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java:64]: skip this if `startFieldBlockOffset == 0`

File: core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java [hunk 12]
```diff
@@ -35,40 +35,56 @@ public String getName()
...
-            sliceOutput.writeInt(fieldBlockOffsets[offsetBase + position] - startFieldBlockOffset);
-        }
         EncoderUtil.encodeNullsAsBits(sliceOutput, block);
+
+        if ((rowBlock.getRowIsNull() == null) != (fieldBlockOffsets == null)) {
+            throw new IllegalArgumentException("When rowIsNull is (non) null then fieldBlockOffsets should be (non) null as well");
+        }
+
+        if (fieldBlockOffsets != null) {
+            int[] positions = new int[positionCount + 1];
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java:63]: `positions` -> `newFieldBlockOffsets`

File: core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlock.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+{
+    @Test
+    public void testWhetherFieldBlockOffsetsIsNullWhenThereIsNoNull()
```
Comment (sopel39) [inline core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlock.java:27]: nit: `testWhetherFieldBlockOffsetsIsNullWhenThereIsNoNull` -> `testFieldBlockOffsetsIsNullWhenThereIsNoNullRow`

File: core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlock.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        Block fieldBlock = new ByteArrayBlock(1, Optional.empty(), new byte[]{10});
+        AbstractRowBlock rowBlock = (RowBlock) RowBlock.fromFieldBlocks(1, Optional.empty(), new Block[] {fieldBlock});
+        // Blocks should discard the offset mask during creation if no values are null
+        assertNull(rowBlock.getFieldBlockOffsets());
+    }
+
+    @Test
+    public void testWhetherFieldBlockOffsetsIsNotNullWhenThereIsNull()
```
Comment (sopel39) [inline core/trino-spi/src/test/java/io/trino/spi/block/TestRowBlock.java:36]: `testWhetherFieldBlockOffsetsIsNotNullWhenThereIsNull` -> `testFieldBlockOffsetsIsNotNullWhenThereIsNullRow`

**Review by sopel39 — APPROVED** [review]
lgtm % comments

**Review by skrzypo987 — APPROVED** [review]

**Review by sopel39 — APPROVED** [review]

**Comment by sopel39** [comment]
small comment

**Comment by sopel39** [comment]
Too small of a change for RN.
